perception/object_merger/src/object_association_merger/node.cpp

```diff
@@ -12,77 +12,195 @@
 // limitations under the License.
 
-#include <sensor_msgs/msg/point_cloud2.hpp>
-#include <sensor_msgs/point_cloud2_iterator.hpp>
+#include "object_association_merger/node.hpp"
 
-#include <tf2/LinearMath/Transform.h>
-#include <tf2/convert.h>
-#include <tf2/transform_datatypes.h>
+#include "object_association_merger/utils/utils.hpp"
+#include "tier4_autoware_utils/tier4_autoware_utils.hpp"
+
+#include <boost/optional.hpp>
 
 #include <chrono>
 #include <unordered_map>
-// #include <tf2_sensor_msgs/msg/tf2_sensor_msgs.hpp>
-#include <object_association_merger/node.hpp>
+
 #define EIGEN_MPL2_ONLY
 #include <Eigen/Core>
 #include <Eigen/Geometry>
 
+using Label = autoware_auto_perception_msgs::msg::ObjectClassification;
+
+namespace
+{
+boost::optional<geometry_msgs::msg::Transform> getTransform(
+  const tf2_ros::Buffer & tf_buffer, const std::string & source_frame_id,
+  const std::string & target_frame_id, const rclcpp::Time & time)
+{
+  try {
+    geometry_msgs::msg::TransformStamped self_transform_stamped;
+    self_transform_stamped = tf_buffer.lookupTransform(
+      /*target*/ target_frame_id, /*src*/ source_frame_id, time,
+      rclcpp::Duration::from_seconds(0.5));
+    return self_transform_stamped.transform;
+  } catch (tf2::TransformException & ex) {
+    RCLCPP_WARN_STREAM(rclcpp::get_logger("object_association_merger"), ex.what());
+    return boost::none;
+  }
+}
+
+bool transformDetectedObjects(
+  const autoware_auto_perception_msgs::msg::DetectedObjects & input_msg,
+  const std::string & target_frame_id, const tf2_ros::Buffer & tf_buffer,
+  autoware_auto_perception_msgs::msg::DetectedObjects & output_msg)
+{
+  output_msg = input_msg;
+
+  /* transform to world coordinate */
+  if (input_msg.header.frame_id != target_frame_id) {
+    output_msg.header.frame_id = target_frame_id;
+    tf2::Transform tf_target2objects_world;
+    tf2::Transform tf_target2objects;
+    tf2::Transform tf_objects_world2objects;
+    {
+      const auto ros_target2objects_world =
+        getTransform(tf_buffer, input_msg.header.frame_id, target_frame_id, input_msg.header.stamp);
+      if (!ros_target2objects_world) {
+        return false;
+      }
+      tf2::fromMsg(*ros_target2objects_world, tf_target2objects_world);
+    }
+    for (auto & object : output_msg.objects) {
+      tf2::fromMsg(object.kinematics.pose_with_covariance.pose, tf_objects_world2objects);
+      tf_target2objects = tf_target2objects_world * tf_objects_world2objects;
+      tf2::toMsg(tf_target2objects, object.kinematics.pose_with_covariance.pose);
+      // TODO(yukkysaito) transform covariance
+    }
+  }
+  return true;
+}
+
+bool isUnknownObjectOverlapped(
+  const autoware_auto_perception_msgs::msg::DetectedObject & unknown_object,
+  const autoware_auto_perception_msgs::msg::DetectedObject & known_object)
+{
+  constexpr double precision_threshold = 0.4;
+  constexpr double recall_threshold = 0.5;
```

Comment:
[Q] I'm not sure about these values, are constants ok? not parameter?

Reply:
I will change it to a parameter. :+1:

Reply:
@miursh fixed
[b381f1c](https://github.com/autowarefoundation/autoware.universe/pull/1001/commits/b381f1cef48efdaf238f2bd13a7a470e7924c7fe)